src/libraries/System.Text.RegularExpressions/gen/UpgradeToRegexGeneratorAnalyzer.cs

```diff
@@ -253,15 +253,10 @@ private static bool IsConstant(IArgumentOperation argument)
         /// <summary>
         /// Detects whether or not the current project is using top-level statements.
         /// </summary>
-        private static async Task<bool> ProjectUsesTopLevelStatements(Compilation compilation, CancellationToken cancellationToken)
+        private static bool ProjectUsesTopLevelStatements(Compilation compilation)
         {
-            SyntaxNode? root = await compilation.SyntaxTrees.FirstOrDefault().GetRootAsync(cancellationToken).ConfigureAwait(false);
-            if (root is null)
-            {
-                return false;
-            }
-
-            return root.DescendantNodesAndSelf().Where(node => node.IsKind(SyntaxKind.GlobalStatement)).Any();
+            var programType = compilation.GetTypeByMetadataName("Program");
+            return programType is not null && !programType.GetMembers("<Main>$").IsEmpty;
         }
```

Comment:
This logic was the one suggested by @sharwell from the Roslyn team. While your approach works as well, I'm hesitant to take it mainly because it depends on the fact that the root type for top-level-statement methods is always called `Program`, and while that may be true today, I'm not sure if it will be true in the future. I recall a recent discussion around this and I believe there was some consideration to probably changing it. For what to do here, I would just do whatever @sharwell and other folks from the Roslyn team suggest.

Reply:
>The top-level syntax tree might not be the first tree.

Can you provide an example when that is not the case?

Reply:
I didn't dig exactly into what order syntax trees will be. But to reproduce in a unit test, just supply multiple source files in the order you want. In reality, the order of syntax trees probably depends on the order of file paths given to the compiler (either the user invoked csc manually or MSBuild does its magic). For both cases I don't think it's reliable to depend on that.

But I'd very much like to hear from @sharwell.

Reply:
Are you speaking of an incremental source generator? This is a regular `DiagnosticAnalyzer`.

Reply:
Ah, sorry, this is an analyzer not a source generator. I believe the new code is the most reasonable option _if_ the condition is retained. I'll leave more comments on the other thread.